app/src/main/java/tech/relaycorp/gateway/data/preference/PublicGatewayPreferences.kt

```diff
@@ -24,11 +27,34 @@ class PublicGatewayPreferences
         preferences.get().getString("address", DEFAULT_ADDRESS)
     }
 
-    suspend fun getAddress() = observeAddress().first()
-    private fun observeAddress() = { address }.toFlow()
+    private fun getAddress(): String = address.get()
     suspend fun setAddress(value: String) = address.setAndCommit(value)
 
-    suspend fun getCogRPCAddress() = COGRPC_ADDRESS
+    fun getCogRPCAddress() = "https://$DEFAULT_ADDRESS"
+
+    @Throws(PublicAddressResolutionException::class)
+    suspend fun resolvePoWebAddress(): ServiceAddress {
```

Comment:
Most of the code in this method is a slight variable of https://github.com/relaycorp/relaynet-courier-android/blob/e5ebacfb91f8590aef0c7726a49dd1719ae572e9/app/src/main/java/tech/relaycorp/courier/data/model/MessageAddress.kt#L44-L61

I think we'll want to move this logic to a shared library in the future, but I didn't want to add it to the core library for two reasons:

- The dependencies of the DoH client are relatively "heavy" and this whole logic isn't needed on the Relaynet-compatible apps.
- DoH is just an implementation detail and Relaynet itself doesn't actually _require_ it. Relaynet only requires DNSSEC on the "unified address" (e.g., `frankfurt.relaycorp.cloud`), and I'm simply using DoH to get DNSSEC working in a quick and safe way.

Reply:
Yeah, they are very similar, we could extract that to a resolver that both classes use. That way it would be easier to mock as well, and test separately.

On the MessageAddress is not as easy to mock, since it's a data model, and not something we inject. But on the resolver extracted class, we could have a global method that does the same thing, to simplify that change.

Reply:
Agreed! I've just created https://github.com/relaycorp/relayverse/issues/20 to track that work